### ISSUE TITLE
PPO Training Reward Collapses After Evaluation Callback

### ISSUE THREAD

**Issue by Oussama-ziadi** [ugr-sail/sinergym#511]
## Environment
- **Sinergym**: 3.9.0
- **Stable Baselines3**: 2.x
- **Gymnasium**: 1.x
- **EnergyPlus**: 23.2.0
- **Algorithm**: PPO
- **OS**: Ubuntu 22.04

Training rewards improve steadily until the first evaluation callback triggers, then **collapse ~30% and never recover**. Using `LoggerEvalCallback` from Sinergym with `NormalizeObservation` wrapper (running statistics mode).

![Training curve showing reward collapse after evaluation]

what I expect from my agent :

1. Training improves (rewards become less negative)
2. Evaluation runs periodically
3. Training resumes and continues improving from where it left off

what I got:

<img width="1073" height="533" alt="Image" src="https://github.com/user-attachments/assets/b9734ee1-d8a5-4d8a-a8fb-c2f49a839e9b" />

1. Training improves: -105k → -102k over ~3 episodes ✓
2. Evaluation triggers at ~157k timesteps ✓
3. Training resumes but rewards crash to -130k and stay there ✗

The policy experiences **catastrophic forgetting** immediately after evaluation and never recovers.

## Root Cause Hypothesis

`LoggerEvalCallback._on_event()` does:
```python
self.train_env.close()      # Close train environment
self._sync_envs()           # Copy stats train → eval (one-way)
# ... evaluation runs ...
self.eval_env.close()
self.train_env.reset()      # Reset train - BUT stats may be lost/reset here!
```

The `NormalizeObservation` wrapper maintains running statistics (`mean`, `var`) that the policy was trained on. After `train_env.reset()`, these statistics appear to be lost or reset, causing:
- Policy receives observations normalized with different statistics
- Actions become nonsensical
- PPO (on-policy) immediately learns from bad experiences
- Death spiral - never recovers

`#Minimal` code reproduction
import gymnasium as gym
from stable_baselines3 import PPO
from sinergym.utils.callbacks import LoggerEvalCallback
from sinergym.utils.wrappers import (
    LoggerWrapper, NormalizeObservation, NormalizeAction, CSVLogger
)

# Training environment
env = gym.make('Eplus-5zone-hot-continuous-v1')
env = NormalizeAction(env)
env = NormalizeObservation(env)  # Running statistics mode
env = LoggerWrapper(env)
env = CSVLogger(env)

# Evaluation environment (identical setup)
eval_env = gym.make('Eplus-5zone-hot-continuous-v1')
eval_env = NormalizeAction(eval_env)
eval_env = NormalizeObservation(eval_env)
eval_env = LoggerWrapper(eval_env)
eval_env = CSVLogger(eval_env)

model = PPO('MlpPolicy', env, verbose=1)

eval_callback = LoggerEvalCallback(
    eval_env=eval_env,
    train_env=env,
    n_eval_episodes=3,
    eval_freq_episodes=3,  # Evaluate every 3 episodes
    deterministic=True,
    verbose=2,
)

# Training will collapse after first evaluation
model.learn(total_timesteps=500_000, `callback=eval_callback)`

## Questions

1. Is `NormalizeObservation` state expected to persist through `env.close()` + `env.reset()` cycle?
2. Should `LoggerEvalCallback._sync_envs()` also restore train_env stats after evaluation?
3. Is there a recommended pattern for mid-training evaluation with stateful wrappers in Sinergym?

**Any guidance on properly handling stateful normalization wrappers with Sinergym's evaluation callbacks would be greatly appreciated!**

**Comment by Oussama-ziadi** [comment]
Hi @AlejandroCN7 , thanks for the quick response
After extensive debugging, I can confirm that `NormalizeObservation` metrics are preserved correctly - that was my initial suspicion but it turned out not to be the issue.

**BEFORE EVAL: mean=[6.526, 15.721, ...], var=[11.888, 77.374, ...]
AFTER EVAL:  mean=[6.526, 15.720, ...], var=[11.888, 77.375, ...]  # Identical!**

The actual problem I discovered is with PPO's internal state `(_last_obs)` becoming desynchronized after `LoggerEvalCallback` resets the training environment.

#My Setup

`env = gym.make(BASE_ENV_ID, **env_kwargs)
env = ActionDeltaConstraintWrapper(env, config=ACTION_DELTA_CONFIG) 
env = NormalizeAction(env) 
env = NormalizeObservation(env)
env = LoggerWrapper(env)
env = CSVLogger(env)

print("\nCreating evaluation environment...")
eval_kwargs = env_kwargs.copy()
eval_kwargs['env_name'] = f"{env_kwargs['env_name']}-PPO-Eval"

eval_env = gym.make(BASE_ENV_ID, **eval_kwargs)
eval_env = ActionDeltaConstraintWrapper(eval_env, config=ACTION_DELTA_CONFIG) 
eval_env = NormalizeAction(eval_env)
eval_env = NormalizeObservation(eval_env)
eval_env = LoggerWrapper(eval_env)
eval_env = CSVLogger(eval_env)
print("Evaluation environment created")`

# PPO model
`model= PPO(
    policy='MlpPolicy',
    env=env,
    learning_rate=1e-4,
    n_steps=4096,
    batch_size=512,
    n_epochs=8,
    gamma=0.99,
    gae_lambda=0.95,
    target_kl=0.01,
    ent_coef=0.01,
    verbose=2
)

callbacks = []
eval_callback =  LoggerEvalCallback(  
        eval_env = eval_env, 
        train_env = env,
        n_eval_episodes=N_EVAL_EPISODES,
        eval_freq_episodes=EVAL_FREQ_EPISODES,
        deterministic=True, 
        verbose=2,

    )
callbacks.append(eval_callback)
callback = CallbackList(callbacks)

model.learn(
        total_timesteps=TOTAL_TIMESTEPS,
        callback=callback,
        log_interval=5,
        tb_log_name="ppo",
        reset_num_timesteps=True,
        progress_bar=False
    )
    
    # Save final model
    
    final_model_path = f"{MODEL_SAVE_PATH}_final.zip"
    model.save(final_model_path)`

## The Problem

Training rewards improve steadily until the first evaluation callback triggers, then **permanently collapse by ~30%** and never recover:

`Episode 1: -194,000
Episode 2: -148,000
Episode 3: -125,000  ← Improving nicely
Episode 4: -105,000
Episode 5: -102,000
[EVALUATION TRIGGERS HERE]
Episode 6: -130,000  ← Sudden drop!
Episode 7: -132,000  ← Never recovers
Episode 8: -131,000`


<img width="1073" height="533" alt="Image" src="https://github.com/user-attachments/assets/982c4d76-41c1-4f61-b6fa-d7bc8c8cec3f" />
